### PR TITLE
[circt-mc] Introduce `circt-mc` (supersedes #4647)

### PR DESCRIPTION
**Pull request by TaoBi22**
Introduces the `circt-mc` tool that was previously drafted in #4647 - since there's a lot of bloat on that thread from development I figured a new PR for actual review made sense (apologies for how long it ended up taking for me to come back to this tool and get it into better shape!). The tool now has a front-end thanks to the `verif` dialect (assert and assume ops are supported, would be great to add support for `LTL` dialect ops especially and I have some ideas for how to do this, but seems a lot for the initial PR!). Needs a few more tests and I'll flag some review points that I'm aware of and would appreciate feedback on below, but I figured getting this available ahead of the model checking ODM made sense.

One big change from the draft PR is that for simplicity's sake that PR was horrible for memory efficiency, as the update function for every single instance of an op was stored separately. That's now fixed, but because I need to encompass every possible signature of update function (and would rather avoid a lot of unnecessary lookups with a big all-purpose function signature) there's some pretty ugly type variant games being played. I'll flag this in a review comment below, but thoughts would be very appreciated. This also means there's a lot of `circt-lec` refactoring also happening, but I would say net-positive refactoring (e.g. a clear map of ops to semantic representations is available now).

There's also some bloat here from having to work around  the fact that the `circt-lec` infrastructure this all builds upon has a separation between the IR itself and the circuit representation - in theory this is very nice, so I've been careful not to violate it, but in practice it's a little bit impractical as the tool grows (and leads to some weird situations e.g. operation names still being used in the part of the code that shouldn't _really_ know about ops). Walking the IR with a case statement that updates the state for each op would be a much simpler way of model checking, instead of storing a vector of things to update - discussing with @fabianschuiki we wondered if maybe the extra layer of separation here might be a problem in the long term, as we continue to need more than the LEC model of just building a standalone Z3 representation of the circuit and moving on. Probably something to sort and simplify separately from this PR though IMO, once there's something upstream that actually has to deal with the added complexity.

There might also be a little bit of massaging necessary to handle some graph region edge cases here, but as I add tests to cover edge cases that should become clearer.

### CODE REVIEW

File: lib/LogicalEquivalence/Circuit.cpp [hunk 1]
```diff
@@ -12,23 +12,166 @@
...
+#include "llvm/ADT/StringRef.h"
 
 #define DEBUG_TYPE "lec-circuit"
 
 using namespace mlir;
 using namespace circt;
 
+/// Populate the table of combinational transforms
+void Solver::Circuit::populateCombTransformTable() {
+  this->combTransformTable.insert(std::pair(
```
[the remark this answers is not in the document]
Comment (TaoBi22) [reply lib/LogicalEquivalence/Circuit.cpp:31]: It would also be nice to move this into another file, since a) it's a lot of code that, while vital, isn't _necessarily_ relevant to how the tool actually works algorithmically and b) a map that represents the semantics of different ops would be pretty nice to have available for other tools IMO

File: lib/LogicalEquivalence/Circuit.cpp [hunk 2]
```diff
@@ -475,3 +712,330 @@ z3::expr Solver::Circuit::boolToBv(const z3::expr &condition) {
...
+  if (!checkState()) {
+    return false;
+  }
+  updateInputs(count, false);
+  runClockNegedge();
+  return checkState();
+}
+
+/// Update combinatorial logic states (to propagate new inputs/reg outputs)
+void Solver::Circuit::applyCombUpdates() {
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply lib/LogicalEquivalence/Circuit.cpp:860]: Could this directly traverse the IR such that you don't need to have that `combTransformTable` and other tables around?

I guess one of the challenges is that the HW module ops have cycles, and that for primary inputs and registers you want to special treatment or potentially feed in a value from the previous iteration. Do you think the whole module-to-Z3 step could be a function like `(Operation *module, Map<Value, z3::Ast> inputs, Map<Value, z3::Ast> currentRegValues) -> (Map<Value, z3::Ast> outputs, Map<Value, z3::Ast> nextRegValues)`?

File: include/circt/LogicalEquivalence/Circuit.h [hunk 3]
```diff
@@ -75,43 +84,148 @@ class Solver::Circuit {
...
+      std::tuple<mlir::Value, mlir::Value, mlir::Value, llvm::StringLiteral>,
+      /*Variadic Ops*/
+      std::tuple<mlir::OperandRange, llvm::StringLiteral>,
+      /*ICmpOp:*/
+      std::tuple<circt::comb::ICmpPredicate, mlir::Value, mlir::Value,
+                 llvm::StringLiteral>,
+      /*ParityOp & ReplicateOp:*/
+      std::tuple<mlir::Value, unsigned int, llvm::StringLiteral>,
+      /*ExtractOp:*/
+      std::tuple<mlir::Value, uint32_t, int, llvm::StringLiteral>>;
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply include/circt/LogicalEquivalence/Circuit.h:130]: I suspect the indirection through `Circuit` is causing this churn here. Might be worthwhile trying to delete that whole thing? Pretty sure that the hundreds of lines of overhead the indirection introduces could boil down to those 200-400 lines of `performX` calls directly inlined into a visitor, and then calling that from an `op.walk([&](Operation *op){ visitor.dispatch(op); });` call. WDYT?

The LogicExporter feels like it really wants to be something that walks the IR, ideally depth-first such that there are no `constrainResults` calls, taking in values for the inputs and current reg state, and producing new Z3 expressions for the outputs and next reg state. That's a bit annoying because of the depth-first search, so alternatively it could have four maps, with entries for the inputs, current reg values, outputs, and next reg values.

File: lib/LogicalEquivalence/Circuit.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (TaoBi22) [reply lib/LogicalEquivalence/Circuit.cpp:31]: @fabianschuiki Agreed, this is just a clunky workaround for not being able to walk directly, but fundamentally is the core of the LEC and MC logic - being able to encode it more elegantly would be great. Would also be lovely to have it reusable elsewhere - a dedicated collection of comb semantics seems pretty likely to end up being useful in other places.

File: include/circt/LogicalEquivalence/Circuit.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (TaoBi22) [reply include/circt/LogicalEquivalence/Circuit.h:130]: Yeah, I think a lot of this is just bulk working around Circuit. Letting the LogicExporter get its hands on the Z3 directly seems way more elegant, and would prune a lot of the bulk out I think (although I do think the Circuit architecture was a nice separation of concerns for LEC. Just tricky when we need to do more with the IR)

File: lib/LogicalEquivalence/Circuit.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (TaoBi22) [reply lib/LogicalEquivalence/Circuit.cpp:860]: Yeah, this function is basically pretending to traverse the IR directly :(

I like that state representation as a map in the function signature a lot though!! Off the top of my head and not having thought about it too much, I think a function with that signature would do the trick, and clean up the logic a lot.

**Comment by TaoBi22** [comment]
Thanks for all the feedback @fabianschuiki - it seems pretty clear that the Circuit infrastructure needs trimming. Think I mentioned this to you before, but that seems like a lot of change to also include in this PR - seems like a little too much area for mistakes. I guess the question is whether to get this upstreamed and then clean up the infrastructure, or have a separate PR first that prunes away all the Circuit infra and refactors circt-lec, then refactor this PR to make it compatible. What are your thoughts?

**Comment by TaoBi22** [comment]
Probably makes sense to close this now that we have a plan to replace this with a new version based on the SMT dialect stuff that @maerhart is currently upstreaming, which should fix all the infrastructure issues this version was tiptoeing around! :tada: :tada:
